### ISSUE TITLE
Support for arbitrary bind mount directories on read-only FS for rootless containers

### ISSUE THREAD

**Issue by olifre** [opencontainers/runc#1671]
Following the idea outlined here:
https://github.com/singularityware/singularity/issues/1207#issuecomment-350137374
it would be possible to allow for arbitrary bind mount points (i.e. also for directories not existing in the container yet, and while keeping it read-only). 

To summarize, the idea would be:
1. Recreate the top-level directory structure of the container in a scratch-directory, for example a subdirectory of the `--root` directory. 
2. Create the additional, not yet existing directories which are meant to be used as additional bind mounts. 
3. Bind-mount all directories over one-by-one. 
4. `chroot` to that directory and remount read-only to not have the user have permissions on `/`.

**Comment by cyphar** [comment]
Is this a duplicate of #1670? Also we already support doing this with `{"readonly": true}` in `runc`. Can you give an example where it doesn't work (#1670 is about having a root which was readonly in the first place, and then telling `runc` to create a new directory on a readonly filesystem -- which obviously won't work)?

**Comment by olifre** [comment]
The idea discussed here also works with fully read-only FS and is essentially a workaround to allow `{"readonly": true}` to be honoured _and_ custom bind mounts to work. 

This works since the `mkdir` is not done in the `root`, but a new read-writable `root` is made in a scratch directory, and all directories from the read-only FS are bind-mounted over.

**Comment by cyphar** [comment]
But then the `/` will not be the same filesystem as the "root" -- so changes to the top-level rootfs won't be propagated. Unless I'm missing something you're proposing something like

```
# where $rootfs is the ro root
mkdir $rootfs/../scratch
for dir in $rootfs/*
do
  name="$(basename "$dir")"
  mkdir "$rootfs/../scratch/$name"
  mount --rbind -o ro "$dir" "$rootfs/../scratch/$name"
  mount -o remount,ro "$rootfs/../scratch"
  pivot_root "$rootfs/../scratch" .
done
```

**Comment by olifre** [comment]
That's exactly what I am suggesting, thanks for summarizing it so nicely!


> But then the / will not be the same filesystem as the "root" -- so changes to the top-level rootfs won't be propagated.

Exactly. Is this an issue?

**Comment by olifre** [comment]
An alternative approach would be to have overlayfs work with user namespaces, but I'm not sure whether something like this is on the Kernel team's schedule at all.

**Comment by AkihiroSuda** [comment]
> An alternative approach would be to have overlayfs work with user namespaces,

You can do with https://github.com/opencontainers/runc/pull/1688

**Comment by DrDaveD** [comment]
@AkihiroSuda does this work on systems other than Ubuntu?  My understanding is that overlayfs in unprivileged namespace only works on Ubuntu which has a non-standard kernel.

**Comment by AkihiroSuda** [comment]
Unfortunately no, it requires Ubuntu-specific kernel patch: http://kernel.ubuntu.com/git/ubuntu/ubuntu-artful.git/commit/fs/overlayfs?h=Ubuntu-4.13.0-25.29&id=0a414bdc3d01f3b61ed86cfe3ce8b63a9240eba7

However, if you are on XFS, I believe you could use `copy_file_range(2)` for copying files with reflink (CoW).
Since it requires to execute the syscall on a bunch of files, the copy operation would be slower than overlay though.

**Comment by cyphar** [comment]
@olifre 

> > But then the / will not be the same filesystem as the "root" -- so changes to the top-level rootfs won't be propagated.
>
> Exactly. Is this an issue?

Personally, I would find it an issue, yes. Especially since the fix is to just create the directory before you try to bind-mount over it. I'm not sure if there's a clever way of doing it that doesn't break `/`.

**Comment by olifre** [comment]
> Especially since the fix is to just create the directory before you try to bind-mount over it. I

Of course, but that does not work for read-only FS. One such dedicated read-only filesystem extremely efficient to distribute software and containers is CVMFS ( https://cernvm.cern.ch/portal/filesystem ). 
Other approaches may include sysadmins exposing containers for consumption by multiple users via a share which is mounted read-only. 

So I think one should not assume that the container root is writable for all usecases. 
Maybe a compromise would be to offer this functionality only if the json states the container is read-only, or the actual FS is read-only?

**Comment by cyphar** [comment]
> Maybe a compromise would be to offer this functionality only if the json states the container is read-only, or the actual FS is read-only?

I was thinking about this too. I think that if the filesystem is already readonly, then the whole "changes aren't propagated" problem isn't as big (though it still exists because you could remount the root as read-write and suddenly some changes won't propagate anymore...). So this might be acceptable if it isn't done for most "readonly" cases. But I'm concerned that it will make an already-quite-complicated part of our codebase (`rootfs_linux.go`) even more complicated.

I will think about it some more. The other issue with doing this is that it increases the amount of magic handling we have that is outside of the OCI specification...

**Comment by olifre** [comment]
> I will think about it some more. The other issue with doing this is that it increases the amount of magic handling we have that is outside of the OCI specification...

Have you reached a conclusion / are open for further discussion? 

My personal feeling is that read-only filesystems for container distribution are getting rather common - CernVM FS is the most efficient solution I know for "unpacked" containers. For other use cases, administrators may offer unpacked containers via a read-only network file system. 

I am not sure how functionality like this can be taken into account in the OCI specification (I did not find the correct place where handling of `/` is discussed). A few other OCI-compatible container runtimes offer this functionality already. (e.g. Singularity) or have it on their to-do-list (e.g. Charliecloud), so it might also be a worthwhile addition to the OCI spec in general, since it is a real-life use case. If you point me to the correct place to ask for adding this to the spec, this would also be fine with me. 

Sadly, my Go experience is close to zero, otherwise I could also offer help in creating a PR to add this functionality.

**Comment by olifre** [comment]
It seems that the need to have this functionality led to other developments such as:
https://github.com/containers/fuse-overlayfs
which seems to fit well with `podman` and probably also `runc`. 

@cyphar I see you've also taken part in discussions on that project, so I wonder if this basically means that the issue discussed here should be solved with tooling such as `fuse-overlayfs` externally / within tools such as Podman instead of integrating that tooling / leveraging that tooling from `runc`?
